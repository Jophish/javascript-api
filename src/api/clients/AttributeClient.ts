import gateway from '../http/gateway';
import Attribute, { IAttribute } from '../dtos/Attribute';

export default class AttributeClient {
  /**
   * Lists all attributes of a given traffic type
   */
  list(ttId: string): Promise<Array<Attribute>> | Promise<Error> {
    if (typeof ttId === 'string') {
      return gateway.get(`/trafficTypes/${ttId}/schema`).then((res: any) => {
        return res.objects.map(e => new Attribute(e));
      });
    } else {
      return gateway.rejectedReq(new Error('You need to provide a Traffic Type ID to get a list of attributes.'));
    }
  };
  /**
   * Creates an attribute
   */
  create(attribute: IAttribute) {
    try {
      const attr = new Attribute(attribute);
      return gateway.put(attr, `/trafficTypes/${attr.trafficTypeId}/schema`).then((res) => {
        return new Attribute(<any> res);
      });     
    } catch (err) {
      return gateway.rejectedReq(err);
    }
  }
  /**
   * Deletes an attribute
   */
  delete(attribute: IAttribute) {
    try {
      const attr = new Attribute(attribute);
      return gateway.del(`/trafficTypes/${attr.trafficTypeId}/schema/${attr.id}`);     
    } catch (err) {
      return gateway.rejectedReq(err);
    }
  }
}
