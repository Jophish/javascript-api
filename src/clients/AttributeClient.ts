import Client from './Client';
import Attribute, { IAttribute } from '../dtos/Attribute';

export default class AttributeClient extends Client {
  /**
   * Lists all attributes of a given traffic type
   */
  list(ttId: string): Promise<Attribute[]> {
    if (typeof ttId === 'string') {
      return this.gateway.get(`/trafficTypes/${ttId}/schema`).then((res: any) => {
        return res.map(e => new Attribute(e));
      });
    } else {
      return <any> this.gateway.rejectedReq(new Error('You need to provide a Traffic Type ID to get a list of attributes.'));
    }
  };
  /**
   * Creates an attribute
   */
  create(attribute: IAttribute): Promise<Attribute> {
    try {
      const attr = new Attribute(attribute);
      return this.gateway.put(attr, `/trafficTypes/${attr.trafficTypeId}/schema`).then((res) => {
        return new Attribute(<any> res);
      });
    } catch (err) {
      return <any> this.gateway.rejectedReq(err);
    }
  }
  /**
   * Deletes an attribute
   */
  delete(attribute: IAttribute): Promise<boolean> {
    try {
      const attr = new Attribute(attribute);
      return this.gateway.del(`/trafficTypes/${attr.trafficTypeId}/schema/${attr.id}`);
    } catch (err) {
      return <any> this.gateway.rejectedReq(err);
    }
  }
}
