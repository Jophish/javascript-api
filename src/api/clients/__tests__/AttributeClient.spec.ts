jest.mock('../../http/gateway');

import AttributeClient from '../AttributeClient';

test('should be a class (function)', () => {
  expect(typeof AttributeClient).toBe('function');
});

test('instances should expose an API', () => {
  const attribute = new AttributeClient();

  expect(attribute.list).toBeDefined();
  expect(attribute.create).toBeDefined();
  expect(attribute.delete).toBeDefined();
});

test('it should be able to list the attributes given a traffic type id, returning a promise', () => {
  const attribute = new AttributeClient();
  const listPromise = attribute.list('user');
  const listPromise2 = attribute.list('machine');  

  expect(listPromise.then).toBeDefined();
  expect(typeof listPromise.then).toBe('function');  

  listPromise.then((res) => {
    expect(res).toMatchSnapshot();
  });
  listPromise2.then((res) => {
    expect(res).toMatchSnapshot();
  });

  return Promise.all([listPromise, listPromise2]);
});

test('it should be able to create an attribute, returning a promise', () => {
  const attribute = new AttributeClient();
  const createPromise = attribute.create({
    id: 'test1',
    trafficTypeId: 'userTT',
    displayName: 'Test 1'
  });
  const createPromise2 = attribute.create({
    id: 'test2',
    trafficTypeId: 'machineTT',
    displayName: 'Test 2'
  });  

  expect(createPromise.then).toBeDefined();
  expect(typeof createPromise.then).toBe('function');  

  createPromise.then((res) => {
    expect(res).toMatchSnapshot();
  });
  createPromise2.then((res) => {
    expect(res).toMatchSnapshot();
  });

  return Promise.all([createPromise, createPromise2]);
});

test('it should be able to delete an attribute, returning a promise', () => {
  const attribute = new AttributeClient();
  const deletePromise = attribute.delete({
    id: 'lname',
    trafficTypeId: 'userTT',
    displayName: 'Test 1'
  });
  const deletePromise2 = attribute.delete({
    id: 'ip',
    trafficTypeId: 'machineTT',
    displayName: 'Test 2'
  }); 

  expect(deletePromise.then).toBeDefined();
  expect(typeof deletePromise.then).toBe('function');  

  deletePromise.then((res) => {
    expect(res).toMatchSnapshot();
  });
  deletePromise2.then((res) => {
    expect(res).toMatchSnapshot();
  });

  return Promise.all([deletePromise, deletePromise2]);
});
