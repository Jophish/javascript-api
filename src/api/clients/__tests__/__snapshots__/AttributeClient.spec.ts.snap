// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should be able to create an attribute, returning a promise 1`] = `
Object {
  "data": Object {
    "displayName": "Test 1",
    "id": "test1",
    "organizationId": "SplitTesting",
    "trafficTypeId": "userTT",
  },
  "response": Object {
    "dataType": "STRING",
    "description": "",
    "displayName": "French name",
    "id": "frname",
    "isSearchable": false,
    "organizationId": "SplitTesting",
    "trafficTypeId": "userTT",
  },
}
`;

exports[`it should be able to create an attribute, returning a promise 2`] = `
Object {
  "data": Object {
    "displayName": "Test 2",
    "id": "test2",
    "organizationId": "SplitTesting",
    "trafficTypeId": "machineTT",
  },
  "response": Object {
    "dataType": "STRING",
    "description": "Name of the machine running the API",
    "displayName": "Machine Name",
    "id": "mname",
    "isSearchable": false,
    "organizationId": "SplitTesting",
    "trafficTypeId": "machineTT",
  },
}
`;

exports[`it should be able to delete an attribute, returning a promise 1`] = `true`;

exports[`it should be able to delete an attribute, returning a promise 2`] = `true`;

exports[`it should be able to list the attributes given a traffic type id, returning a promise 1`] = `
Object {
  "objects": Array [
    Object {
      "dataType": "STRING",
      "description": "First name",
      "displayName": "Name",
      "id": "1",
      "trafficTypeId": "1",
    },
    Object {
      "dataType": "STRING",
      "description": "Last name",
      "displayName": "LName",
      "id": "2",
      "trafficTypeId": "1",
    },
  ],
}
`;

exports[`it should be able to list the attributes given a traffic type id, returning a promise 2`] = `
Object {
  "objects": Array [
    Object {
      "dataType": "STRING",
      "description": "Name of the machine",
      "displayName": "MachineName",
      "id": "1",
      "trafficTypeId": "2",
    },
    Object {
      "dataType": "STRING",
      "description": "IP of the machine",
      "displayName": "MachineIP",
      "id": "2",
      "trafficTypeId": "2",
    },
  ],
}
`;
